### PR TITLE
Implement a Python PEG parser in Rust

### PR DESCRIPTION
**Pull request by zsol**
## Summary

This massive PR implements an alternative Python parser that will allow LibCST to parse Python 3.10's new grammar features. The parser is implemented in Rust, but it's turned off by default through the `LIBCST_PARSER_TYPE` environment variable. Set it to `native` to enable. The PR also enables new CI steps that test just the Rust parser, as well as steps that produce binary wheels for a variety of CPython versions and platforms.

Note: this PR aims to be roughly feature-equivalent to the main branch, so it doesn't include new 3.10 syntax features. That will be addressed as a follow-up PR.

The new parser is implemented in the `native/` directory, and is organized into two rust crates: `libcst_derive` contains some macros to facilitate various features of CST nodes, and `libcst` contains the `parser` itself (including the Python grammar), a `tokenizer` implementation by @bgw, and a very basic representation of CST `nodes`. Parsing is done by
1. **tokenizing** the input utf-8 string (bytes are not supported at the Rust layer, they are converted to utf-8 strings by the python wrapper)
2. running the **PEG parser** on the tokenized input, which also captures certain anchor tokens in the resulting syntax tree
3. using the anchor tokens to **inflate** the syntax tree into a proper CST

These steps are wrapped into a high-level `parse_module` API [here](https://github.com/zsol/LibCST/blob/parser/native/libcst/src/lib.rs#L43), along with `parse_statement` and `parse_expression` functions which all just accept the input string and an optional encoding.

These Rust functions are exposed to Python [here](https://github.com/zsol/LibCST/blob/parser/native/libcst/src/py.rs) using the excellent [PyO3](https://pyo3.rs/) library, plus an `IntoPy` trait which is mostly implemented via a macro in `libcst_derive`.

## Test Plan

Apart from Rust unit tests and roundtrip tests, the new implementation is also tested by running the pure python tests on them (by setting `LIBCST_PARSER_TYPE=native` before running `python -m unittest`). All of these are part of CI steps.

## Known missing features

Apart from the soon-to-come 3.10 syntactic features mentioned above, here's an incomplete list of the features the Rust implementation lacks compared to the pure Python parser:

* No support for limiting the parser to a certain Python version. The parser will accept input that's valid in any of Python 3.9, 3.8, 3.7, or 3.6 at the moment.
* The native parser API only accepts utf-8 encoded strings. Support for bytes is provided in the wrapping python layer.
* In some places the old parser returned a tuple of CSTNodes, the new parser returns a list of them (see adjusted unit tests)

### CODE REVIEW

**Comment by bgw** [comment]
Oh my god. The mad lad, @zsol actually did it. Congratulations!

I don't think I can reasonably read through all this code, but I might look through it a bit later. None of that should prevent you from merging though.

**Comment by zsol** [comment]
Thanks! I'd still appreciate any review or thoughts for even parts of the code, because I'm 100% sure there are some questionable decisions in there that would benefit from ... questioning :) Can incorporate any changes in future PRs.

**Comment by stroxler** [comment]
Wow, there's a lot here! I'll also try to read through parts of it.

My view is that if tests pass we can probably go ahead and merge it even if there are questionable bits as long as the CST data itself is reasonable, nothing prevents us from revisiting the code later and getting 3.10 PEG support landed is pretty time-sensitive.

Before I try to dive in, a few high level questions:

(1) are there any particular parts of the code I should focus on for review? I'm coming off of making some changes to the cpython parser for PEP 677 so I can try to focus on that, but there's a big difference between making one change, even a nontrivial one, versus trying to take in the whole thing at once!

(2) are there any changes you'd like to make on top of this? It might help me build context to not just read the code but make a small modification or two if there's something you already have in mind.

(3) for code that can parse using either the old or new parser, is the produced CST the same? My impression is yes given your description but just want to double-check. In my mind even if there are messy bits, if this is true we should err on the side of merging fast, especially since it's off by default.

**Comment by zsol** [comment]
> are there any particular parts of the code I should focus on for review? I'm coming off of making some changes to the cpython parser for PEP 677 so I can try to focus on that, but there's a big difference between making one change, even a nontrivial one, versus trying to take in the whole thing at once!

In that case, I'd recommend `grammar.rs` which implements step 2 from the PR description. Specifically I'm a bit worried about accepting syntax that shouldn't be accepted.

> are there any changes you'd like to make on top of this? It might help me build context to not just read the code but make a small modification or two if there's something you already have in mind.

Yeah, I'd like to add match statements myself, but you could try extending the grammar to support parenthesized context managers (https://bugs.python.org/issue12782)

> for code that can parse using either the old or new parser, is the produced CST the same?

Yes! The only case where this isn't true is the tuple-vs-list problem I mention in the PR description. PyO3 automatically converts rust's `List<_>` into python `List[...]`, but in some cases the pure python parser produces a tuple.

File: libcst/_nodes/tests/test_import.py
Comment (thatch) [inline libcst/_nodes/tests/test_import.py:403]: Curious if this is intentional -- the old syntax w/o parens is valid.  Does the new parser reject it?

File: libcst/_nodes/tests/test_atom.py
Comment (thatch) [inline libcst/_nodes/tests/test_atom.py:1125]: Do you have a plan for making them work?  This is an important part of the API.

File: libcst/_parser/detect_config.py
Comment (thatch) [inline libcst/_parser/detect_config.py:114]: I'm not 100% sure this has good tests.  Does this avoid having to tokenize twice?

**Comment by thatch** [comment]
Can you expand more on

> No support for limiting the parser to a certain Python version

Are you intending on adding that before a non-pre release?  I put a decent chunk of effort in supporting old versions and even have a draft going back to 2.3 that I'd be happy to rework if version-dependent support existed. See https://github.com/Instagram/LibCST/pull/275.

File: libcst/_nodes/tests/test_atom.py
Comment (zsol) [reply libcst/_nodes/tests/test_atom.py:1125]: Yeah it's possible to make it work, but my priority is to get 3.10 working first. The old parser won't be removed until this feature is supported by the new one.

File: libcst/_parser/detect_config.py
Comment (zsol) [reply libcst/_parser/detect_config.py:114]: Hmm, I don't think this code is ever getting run in native mode. Will probably have to be reimplemented on the rust side in the future. As far as I can tell this is only being used to figure out the exact grammar version to be used.

**Comment by stroxler** [comment]
@thatch my impression of what @zsol was saying about python versions is just that the parser won't throw errors if you use Python 3.9 syntax in a 3.6 codebase.

In my mind this isn't a huge problem, generally there are other ways to get pretty fast feedback about syntax errors, and it's not critical that LibCST deal with it... if there's an exception to that I'd say it's actually on the *output* of codemods, not the input, but presumably that's not the parser anyway

File: native/libcst/src/parser/grammar.rs
Comment (stroxler) [inline native/libcst/src/parser/grammar.rs:773]: [nit / curiosity] The official grammar uses `star_named_expression ', ' [star_named_expressions]` here.

This code is needed to handle empty tuples... I'm actually having trouble figuring out where the official grammar does that.

File: native/libcst/src/parser/grammar.rs
Comment (stroxler) [inline native/libcst/src/parser/grammar.rs:796]: [compat] for 3.11 we'll need except_star_block handling. Makes sense not to have yet

File: native/libcst/src/parser/grammar.rs
Comment (stroxler) [inline native/libcst/src/parser/grammar.rs:835]: [nit] most of the time we have helper rules after, this one is before.

File: native/libcst/src/parser/grammar.rs
Comment (stroxler) [inline native/libcst/src/parser/grammar.rs:973]: [compat] need to add parenthesized with statements, which also allow trailing commas (later)

File: native/libcst/src/parser/grammar.rs
Comment (stroxler) [inline native/libcst/src/parser/grammar.rs:984]: [compat] needs to include `)` in the peek ahead once we do parenthesized with

File: native/libcst/src/parser/grammar.rs
Comment (stroxler) [inline native/libcst/src/parser/grammar.rs:996]: [compat] this is the start of another section where our ordering diverges wildly from the official grammar

File: native/libcst/src/parser/grammar.rs
Comment (stroxler) [inline native/libcst/src/parser/grammar.rs:1001]: [nit] it sems like in most cases where the official grammar has `','.something+`, we've been able to use a similar form involving comma-separated stuff. Does that not apply here?

I think what we have is equivalent but the grammar diverges

File: native/libcst/src/parser/grammar.rs
Comment (stroxler) [inline native/libcst/src/parser/grammar.rs:1006]: [compat] (same as for global)

File: native/libcst/src/parser/grammar.rs
Comment (stroxler) [inline native/libcst/src/parser/grammar.rs:1018]: [nit / curiosity] this diverges from the grammar because we have separate rules for del_target and del_targets

Is that a historical thing, e.g. we wrap one in a tuple and not the other so it's easier to have two handler actions?

**Comment by bgw** [comment]
> the parser won't throw errors if you use Python 3.9 syntax in a 3.6 codebase

How does this work for bits of Python grammar that were changed in backwards-incompatible ways? In older versions of Python, it was valid to use `async` and `await` as variable names. In 3.7+, it's no longer valid. However, if you enable the `async`/`await` tokenizer hacks on 3.7+ code, you might run into issues with `async` generator expressions (https://bugs.python.org/issue31708).

File: native/libcst/src/parser/grammar.rs
Comment (stroxler) [inline native/libcst/src/parser/grammar.rs:1063]: [nit / curiosity] another case where grammar just uses a `','.dotted_as_name+` and we don't, it seems like we could just use a comma separated rule here

File: native/libcst/src/parser/grammar.rs
Comment (stroxler) [inline native/libcst/src/parser/grammar.rs:1073]: [compat] this rule diverges pretty heavily from the latest grammar which uses a recursive rule instead.

I'm pretty certain they are equivalent

**Review by stroxler — APPROVED** [review]
I read the grammar itself (not the entire grammar.rs file, just up to line 1180 or so) end-to-end and compared to the main grammar at
https://docs.python.org/3.11/reference/grammar.html
as well as, where necessary, the full cpython PEG at
https://github.com/python/cpython/blob/main/Grammar/python.gram

Overall it looks good to me, especially in light of having test coverage.

I flagged a number of things:

- [substantive] I think I only found 2 or 3 things that might be wrong, and
   none of them seem like they ought to block merge of an off-by-default
   new parser - I'd say address them if you want or we can merge and follow
   up later

- [nit / curiousity] a few places the rules seemed awkward, or slightly out of
   sync with the python official grammar in ways that appeared unnecessary.
   I definitely may be missing something though, marked these as [nit], [curiosity]
   or [both] and it's probably fine to merge without addressing (although if there
   are reasons I'd appreciate an explanation since I'd learn something!)

- [compat] I've marked rules that are not in a great order or otherwise have
   cosmetic issues relative to the official grammar as [compat]. Feel free to
   ignore these, they shouldn't block land but I do think I might follow up
   later with a reorder because that would make checking them against
   one another easier going forward

**Comment by stroxler** [comment]
> How does this work for bits of Python grammar that were changed in backwards-incompatible ways? 

Not sure - I was mostly describing my understanding of @zsol's comment about supporting all versions, but it does appear to me that the new parser is using `async` / `await` as dedicated tokens.

I wonder if a solution is to only use the new parser, when we roll it out as default, only for 3.7+? The main motivation to move to the PEG is to properly support newer python versions. If we want to support 3.6 properly with the native parser we could always put the time into that later.

File: native/libcst/src/parser/grammar.rs
Comment (zsol) [reply native/libcst/src/parser/grammar.rs:42]: It's part of the trait, you need to specify an associated type that will be used in the peg implementation: https://docs.rs/peg/0.7.0/peg/trait.Parse.html

File: native/libcst/src/parser/grammar.rs
Comment (zsol) [reply native/libcst/src/parser/grammar.rs:164]: I was surprised at this as well, but apparently it doesn't break any tests. I wasn't able to figure out what this is used for in CPython. Maybe an optimization?

File: native/libcst/src/parser/grammar.rs
Comment (zsol) [reply native/libcst/src/parser/grammar.rs:232]: I pulled out `_conditional_expression` because it's also used for parsing f-strings. The official grammar doesn't include a rule for f-strings, they're parsed in a fairly ad-hoc way. Arguably this implementation is "more correct" for some definition of correct.

File: native/libcst/src/parser/grammar.rs
Comment (zsol) [reply native/libcst/src/parser/grammar.rs:358]: I'll add a note here. This is to avoid bloating the parser cache.

File: native/libcst/src/parser/grammar.rs
Comment (zsol) [reply native/libcst/src/parser/grammar.rs:680]: Same reason as `_conditional_expression` above. It's used for f-string parsing.

File: native/libcst/src/parser/grammar.rs
Comment (zsol) [reply native/libcst/src/parser/grammar.rs:1001]: The official grammar has the luxury of discarding all those separators. We need to refer to them later, which is why this is a bit more complex. Eventually I cooked up the `separated` and `separated_trailer` helper meta-rules, but
1. I'm not sure they improve readability
2. I was too lazy to go back and refactor every occurrence of this pattern. This is tech debt TODO

File: native/libcst/src/parser/grammar.rs
Comment (zsol) [reply native/libcst/src/parser/grammar.rs:1073]: I forget why this was necessary. Maybe it wasn't? Not sure.

File: native/libcst/src/parser/grammar.rs
Comment (zsol) [reply native/libcst/src/parser/grammar.rs:1018]: Yeah this change is because I needed to be 100% backwards compatible with the pure python parser, which produces a tuple in one case and not the other.

File: native/libcst/src/parser/grammar.rs
Comment (zsol) [reply native/libcst/src/parser/grammar.rs:706]: Yes, this is just for convenience.

File: native/libcst/src/parser/grammar.rs
Comment (zsol) [reply native/libcst/src/parser/grammar.rs:633]: Yeah that's a 3.10 thing. FWIW I accidentally checked in the grammar I based the parser on :) https://github.com/zsol/LibCST/blob/parser/native/libcst/Grammar

File: native/libcst/src/parser/grammar.rs
Comment (zsol) [reply native/libcst/src/parser/grammar.rs:99]: This is to preserve LibCST behavior: https://github.com/Instagram/LibCST/blob/45c0d96c06a8f20e330e6afd671ff7145a3bab8e/libcst/_parser/conversions/expression.py#L157

File: native/libcst/src/parser/grammar.rs
Comment (zsol) [reply native/libcst/src/parser/grammar.rs:773]: I remember being confused about this, but I don't remember if I found how this works in CPython. Maybe I just gave up and implemented it **MY WAY** :stuck_out_tongue_closed_eyes:

**Comment by jreese** [comment]
> No support for limiting the parser to a certain Python version. The parser will accept input that's valid in any of Python 3.9, 3.8, 3.7, or 3.6 at the moment.

@zsol can you confirm if you intend to add support for this functionality before the Rust parser becomes enabled by default? ie, have separate grammar for each major release of CPython, and ensure strict parsing when given a target version? 

There is a lot of work that Tim had in-flight for supporting older releases of CPython, potentially back to 2.x versions, that I don't want to lose, as well as the functionality in guaranteeing correctness and safety when parsing/modifying/dumping trees for production code. Eg, when writing codemods, we want to know that if we ask to parse with 3.7 code, that we're not getting trees/nodes/grammar that are only available in 3.8 or 3.9, and vice versa.

**Comment by zsol** [comment]
Yeah this feature won't be lost one way or another. Having said that, I
don't think supporting python 2 grammars is very valuable at this point, as
I commented on Tim's issue recently

**Pull request merged**
